### PR TITLE
HIVE-29637: Incorrect Results for IS_NULL Predicate on Partition Column

### PR DESCRIPTION
**Pull request by deniskuzZ**
…n Column

### What changes were proposed in this pull request?

IcebergTableUtil.makeSpecFromName() — a new helper that builds a Hive-compatible partition spec map from Iceberg's PartitionData. When a partition field value is null, it maps it to an actual NULL instead of the literal string "null".

### Why are the changes needed?

 Queries with WHERE <partition_col> IS NULL on Iceberg tables returned incorrect results (empty result set).

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

mvn test -Dtest=TestIcebergCliDriver -Dqfile=iceberg_isnull_partition_pruning.q

### CODE REVIEW

**Review by aturoczy — APPROVED** [review]

**Review by rubenada — APPROVED** [review]

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 1]
```diff
@@ -2316,8 +2316,8 @@ public List<Partition> getPartitionsByExpr(org.apache.hadoop.hive.ql.metadata.Ta
             PartitionData partitionData = IcebergTableUtil.toPartitionData(task.partition(), spec.partitionType());
             String partName = spec.partitionToPath(partitionData);
 
-            Map<String, String> partSpecMap = Maps.newLinkedHashMap();
-            Warehouse.makeSpecFromName(partSpecMap, new Path(partName), null);
```
Comment (zabetak) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:2320]: There are more calls to `Warehouse#makeSpecFromName` methods in `IcebergTableUtil#convertNameToMetastorePartition`, `IcebergQueryCompactor`, and potentially other places as well. Do we need to update them as well?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 2]
```diff
@@ -452,6 +452,25 @@ public static void performMetadataDelete(Table icebergTable, String branchName,
     deleteFiles.deleteFromRowFilter(exp).commit();
   }
 
+  /**
+   * Parses an Iceberg partition path into a Hive-compatible spec map.
```
Comment (zabetak) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: The method somewhat "overrides" the default behavior of `Warehouse.makeSpecFromName` but it's unclear why this specificity is required for Iceberg especially since we want to create a "Hive-compatible" spec map.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 3]
```diff
@@ -452,6 +452,25 @@ public static void performMetadataDelete(Table icebergTable, String branchName,
...
+   */
+  public static Map<String, String> makeSpecFromName(String partName,
+      PartitionSpec spec, PartitionData data) {
+    Map<String, String> partSpecMap = Maps.newLinkedHashMap();
+    Warehouse.makeSpecFromName(partSpecMap, new Path(partName), null);
+
+    List<PartitionField> fields = spec.fields();
+    for (int i = 0; i < fields.size(); i++) {
+      if (data.get(i, Object.class) == null) {
+        partSpecMap.put(fields.get(i).name(), null);
```
Comment (zabetak) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: If we can alter the `partSpecMap` based on the `PartitionSpec` then why do we even need to pass through the `Warehouse.makeSpecForName` method? Can we create the spec map exclusively using the Iceberg APIs?

File: iceberg/iceberg-handler/src/test/queries/positive/iceberg_isnull_partition_pruning.q [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  int_col INT,
+  bigint_col BIGINT,
+  boolean_col BOOLEAN
+) PARTITIONED BY (str_col String, tinyint_col int)
+STORED BY iceberg;
+
+INSERT INTO ice01 PARTITION (str_col, tinyint_col)
+  SELECT index, date_col, timestamp_col, string_col, double_col, float_col, decimal_col, smallint_col, int_col, bigint_col, boolean_col, str_col, tinyint_col FROM sample01;
+
+SELECT * FROM ice01 WHERE str_col is NULL;
```
Comment (zabetak) [inline iceberg/iceberg-handler/src/test/queries/positive/iceberg_isnull_partition_pruning.q]: Should I add tests for partition with nulls on other type of columns?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 5]
```diff
@@ -452,6 +452,25 @@ public static void performMetadataDelete(Table icebergTable, String branchName,
     deleteFiles.deleteFromRowFilter(exp).commit();
   }
 
+  /**
+   * Parses an Iceberg partition path into a Hive-compatible spec map.
+   * Unlike {@link Warehouse#makeSpecFromName}, this correctly represents null partition values
+   * as {@code null} instead of the literal string "null".
```
Comment (zabetak) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: Based on the comment I get the impression that for other table formats where we use the string "null" things work fine so I am wondering why this is only a problem for Iceberg?

**Review by zabetak — COMMENTED** [review]
Left some high level comments/questions cause I am not very familiar with the modified APIs.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 1 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:2320]: The other call sites (e.g. IcebergTableUtil#convertNameToMetastorePartition) don't have access to `partitionData` — they only receive the partition path string. Without the underlying partition data, there's no way to disambiguate whether `"null"` in the path originated from a true `NULL` value or from a literal string `"null"`. The fix here works precisely because we have `partitionData` available to check the actual value. If those other paths turn out to be affected, they'd need a different approach to obtain that context.

**Note**: IcebergQueryCompactor should be fine — it deals with compaction where `NULL` filtering semantics aren't in play

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 2 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: The issue is in how the partition path string is generated and then parsed back. `spec.partitionToPath(partitionData)` uses Iceberg's toHumanString() to convert partition values to path segments. For `NULL` values, toHumanString() produces the string `"null"`. So the path looks like `col=null`. When `Warehouse.makeSpecFromName` parses this path back into a spec map, it just sees the literal string `"null"` as the value — it has no way to know whether this came from an actual `NULL` or from a legitimate string value `"null"`. It puts `"null"` (the string) into the spec map rather than `NULL`.

The fix uses `partitionData` to check whether the value is actually `NULL` at the Iceberg level, and if so, puts `NULL` into the spec map directly. We need this because toHumanString() is lossy — it  conflates `NULL` and the string `"null"` into the same path representation.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 3 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: we can't build the spec map exclusively from Iceberg APIs

File: iceberg/iceberg-handler/src/test/queries/positive/iceberg_isnull_partition_pruning.q (hunk 4 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/test/queries/positive/iceberg_isnull_partition_pruning.q]: testing one column type is sufficient to verify the logic. Additional type-specific tests would be redundant here.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 5 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: explained in https://github.com/apache/hive/pull/6515#discussion_r3337070347

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 2 above)
Comment (zabetak) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: Thanks for the additional details. I have two small follow-up questions:
1. How do we avoid this problem in non-Iceberg tables? 
2. Do we have test coverage for partitioning where data contain both the `"null"` string and actual `NULL`?

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: 1.  this is an Iceberg-specific problem.
2. added

Comment (zabetak) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: For non-Iceberg tables I assume that actual NULL values go into the default partition (`hive.exec.default.partition.name`). I am trying to understand if when we built the spec we should have the value of `hive.exec.default.partition.name` or `null` and if it makes any difference.

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: Mapping null to __HIVE_DEFAULT_PARTITION__ didn't work in PCR and filter was replaced with false.
unlike prunePartitionNames
````
  if (partitionValue.equals(defaultPartitionName)) {
    convertedValues.add(null); // Null for default partition.
  }  
````
evalExprWithPart doesn't have null-awareness

Comment (zabetak) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: I assume that you refer to `org.apache.hadoop.hive.ql.optimizer.ppr.PartExprEvalUtils#evalExprWithPart`. If the latter does not have null-awareness then it means that the wrong result issue affects all partitioned tables and not only Iceberg.

I tried the following test case on master and it seems to confirm my hypothesis.

```sql
create table pcr_t1 (key string, value string) partitioned by (ds string);
INSERT INTO pcr_t1 PARTITION (ds) SELECT 'A', 'V1', '2000-04-08' ;
INSERT INTO pcr_t1 PARTITION (ds) SELECT 'B', 'V2', 'null';
INSERT INTO pcr_t1 PARTITION (ds) SELECT 'C', 'V3', null;
select key, value, ds from pcr_t1 where ds is null;
```

Should we treat this as apart?

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: fixed in e8c67d57ff6bf9525cc0cb7456db2612e4c9901e, @zabetak please re-check

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartExprEvalUtils.java [hunk 6]
```diff
@@ -82,9 +87,15 @@ static public Object evalExprWithPart(ExprNodeDesc expr, Partition p) throws Hiv
...
-          .convert(entry.getValue()));
+
+      String partitionValue = entry.getValue();
+      if (partitionValue.equals(defaultPartitionName)) {
+        partValues.add(null); // Null for default partition.
+      } else {
+        partValues.add(ObjectInspectorConverters.getConverter(
+            PrimitiveObjectInspectorFactory.javaStringObjectInspector, oi)
+            .convert(partitionValue));
+      }
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartExprEvalUtils.java:98]: The description of `hive.exec.default.partition.name` property says the following:
``` 
"The default partition name in case the dynamic partition column value is null/empty string or any other values that cannot be escaped."
```
This code assumes that the value is null but according to the description it may be other things as well.

The change here will fix the `IS [NOT] NULL` predicate but may potentially change the behavior of some queries:
```sql
SELECT key, value, ds FROM pcr_t1 WHERE ds > 'A';
SELECT key, value, ds FROM pcr_t1 WHERE ds < 'A';
```
Do the queries return the same results before/after the changes in the PR for Iceberg and non-Iceberg tables ?
Which behavior should prevail?

File: ql/src/test/queries/clientpositive/pcr_null_partition.q [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+set hive.fetch.task.conversion=none;
+
+drop table if exists pcr_t1;
+create table pcr_t1 (key string, value string) partitioned by (ds string);
+
+insert into pcr_t1 partition (ds) select 'A', 'V1', '2000-04-08';
+insert into pcr_t1 partition (ds) select 'B', 'V2', 'null';
+insert into pcr_t1 partition (ds) select 'C', 'V3', null;
+
+explain select key, value, ds from pcr_t1 where ds is null;
```
Comment (zabetak) [inline ql/src/test/queries/clientpositive/pcr_null_partition.q:11]: Is there something useful in the EXPLAIN output? If not we can drop those.

File: ql/src/test/queries/clientpositive/pcr_null_partition.q [hunk 8]
```diff
@@ -0,0 +1,17 @@
+set hive.exec.dynamic.partition.mode=nonstrict;
+set hive.fetch.task.conversion=none;
+
+drop table if exists pcr_t1;
+create table pcr_t1 (key string, value string) partitioned by (ds string);
+
+insert into pcr_t1 partition (ds) select 'A', 'V1', '2000-04-08';
+insert into pcr_t1 partition (ds) select 'B', 'V2', 'null';
+insert into pcr_t1 partition (ds) select 'C', 'V3', null;
```
Comment (zabetak) [inline ql/src/test/queries/clientpositive/pcr_null_partition.q:9]: The empty string value is another edge case that may create problems
```sql
insert into pcr_t1 partition (ds) select 'D', 'V4', '';
```

File: ql/src/test/queries/clientpositive/pcr_null_partition.q [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+explain select key, value, ds from pcr_t1 where ds is null;
+select key, value, ds from pcr_t1 where ds is null;
+
+explain select key, value, ds from pcr_t1 where ds is not null;
+select key, value, ds from pcr_t1 where ds is not null order by key;
+
+select key, value, ds from pcr_t1 where ds = 'null';
```
Comment (zabetak) [inline ql/src/test/queries/clientpositive/pcr_null_partition.q:17]: Consider enriching the tests with:
```sql
SELECT key, value, ds FROM pcr_t1 WHERE ds > 'A';
SELECT key, value, ds FROM pcr_t1 WHERE ds < 'A';
```

**Review by zabetak — APPROVED** [review]
Tentatively approve assuming that we all agree that the new behavior is the one desired by all users and the old one is merely a bug.

File: ql/src/test/queries/clientpositive/pcr_null_partition.q (hunk 7 above)
Comment (deniskuzZ) [reply ql/src/test/queries/clientpositive/pcr_null_partition.q:11]: yes, `filterExpr: ds is not null (type: boolean)`

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartExprEvalUtils.java (hunk 6 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartExprEvalUtils.java:98]:
````
Non-Iceberg pcr_t1

  (NULL and '' both land in ds=__HIVE_DEFAULT_PARTITION__)
  
  ┌────────────────┬─────────────────────────────────────────┬─────────────────────────────────────────┬─────────────────┐
  │     Query      │               WITHOUT fix               │                WITH fix                 │ Changed by fix? │
  ├────────────────┼─────────────────────────────────────────┼─────────────────────────────────────────┼─────────────────┤
  │ ds is null     │ (none) ❌                                │ C, D → ds=NULL                          │ ✅  fixed        │
  ├────────────────┼─────────────────────────────────────────┼─────────────────────────────────────────┼─────────────────┤
  │ ds is not null │ A, B                                    │ A, B                                    │ no              │
  ├────────────────┼─────────────────────────────────────────┼─────────────────────────────────────────┼─────────────────┤
  │ ds = 'null'    │ B                                       │ B                                       │ no              │
  ├────────────────┼─────────────────────────────────────────┼─────────────────────────────────────────┼─────────────────┤
  │ ds > 'A'       │ C, D, B → ds=__HIVE_DEFAULT_PARTITION__ │ C, D, B → ds=__HIVE_DEFAULT_PARTITION__ │ no              │
  ├────────────────┼─────────────────────────────────────────┼─────────────────────────────────────────┼─────────────────┤
  │ ds < 'A'       │ A                                       │ A                                       │ no              │
  └────────────────┴─────────────────────────────────────────┴─────────────────────────────────────────┴─────────────────┘
  
  Iceberg ice_01

  (NULL → null partition; '' → real ds='')

  ┌────────────────┬─────────────┬──────────┬─────────────────┐
  │     Query      │ WITHOUT fix │ WITH fix │ Changed by fix? │
  ├────────────────┼─────────────┼──────────┼─────────────────┤
  │ ds is null     │ (none) ❌    │ C        │ ✅  fixed        │
  ├────────────────┼─────────────┼──────────┼─────────────────┤
  │ ds is not null │ A, B, D     │ A, B, D  │ no              │
  ├────────────────┼─────────────┼──────────┼─────────────────┤
  │ ds = 'null'    │ B           │ B        │ no              │
  ├────────────────┼─────────────┼──────────┼─────────────────┤
  │ ds > 'A'       │ B           │ B        │ no              │
  ├────────────────┼─────────────┼──────────┼─────────────────┤
  │ ds < 'A'       │ D, A        │ A, D     │ no (same set)   │
  └────────────────┴─────────────┴──────────┴─────────────────┘

````
 - The fix changes exactly one query on each table type: ds is null (previously returned nothing — the bug)
 -  Range queries (>, <) are unchanged before/after on both table types → no regression.
 - Iceberg is fully SQL-correct throughout because it never conflates NULL with ''.
  
  PR scope is NULL semantics only.

Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartExprEvalUtils.java:98]: For the non-Iceberg case, there is still a bug in the NULL semantics:
```sql
SELECT key, value, ds FROM pcr_t1 WHERE ds > 'A';
```
The query returns C, D, B but for C the ds is `null` and `null > 'A'` should evaluate to `UNKNOWN` and thus `false`. If we consider that empty string `''` is also `null` then the query should only return the `B` value as it happens for Iceberg.

There is a bit of inconsistency among the various predicates but I guess we can tackle this in another JIRA ticket assuming that this PR strictly target `IS NULL` predicate.
